src/databricks/labs/ucx/framework/crawlers.py

```diff
@@ -21,13 +24,22 @@ class DataclassInstance(Protocol):
 
 
 class CrawlerBase(ABC, Generic[Result]):
-    def __init__(self, backend: SqlBackend, catalog: str, schema: str, table: str, klass: type[Result]):
+    def __init__(
+        self,
+        backend: SqlBackend,
+        catalog: str,
+        schema: str,
+        table: str,
+        klass: type[Result],
+        history_log: HistoryLog | None = None,
```

Comment:
**`... | None` constructor-time dependency in a singleton is an anti-pattern and it has led to undefined behaviors before.** 

If we look at things high level, the dependency diagram looks like

```mermaid
flowchart TD
    history -->|optional!!!| crawler
    crawler --> history_appender
    history --> history_appender
    crawler --> snapshot
    history_appender --> append_snapshot
    append_snapshot --> snapshot
```

and it causes some suboptimal assumptions related to `migration_status` table expressed in https://github.com/databrickslabs/ucx/issues/2600#issuecomment-2383202696.

This can be avoided if dependencies are built in a more explicit way:

```
    @job_task
    def crawl_tables(self, ctx: RuntimeContext) -> None:
        ctx.history_log.update_snapshot(ctx.tables_crawler)

    @job_task
    def crawl_udfs(self, ctx: RuntimeContext) -> None:
       ctx.history_log.update_snapshot(ctx.udfs_crawler)
```

and then `update_snapshot` being something like with clear and non-optional dependencies

```mermaid
flowchart TD
    history --> update_snapshot
    crawler --> update_snapshot

    update_snapshot -.-> _appender
    crawler -.-> _appender
    history -.-> _appender
    crawler -.-> snapshot
    snapshot -.-> _appender

```

```
def update_snapshot(self, crawler: CrawlerBase[T]):
    appender = self._create_appender(crawler.get_klass())
    appender.persist(crawler.snapshot(force_refresh=True))
```

this will fit better with things like `migration_status` and `dfsa_*`

Reply:
The implementation has been replaced to fit with this style: instead of being a part of the crawlers, the history mechanism will operate as a wrapper around a select few and controlled explicitly by the relevant workflow tasks.